### PR TITLE
[3.5] Bug 1422376: Fix resolving ImageStreamImage latest tag

### PR DESCRIPTION
**Pull request by mfojtik**
Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1422376
Master: https://github.com/openshift/origin/pull/13089

### CODE REVIEW

File: pkg/image/api/helper.go [hunk 1]
```diff
@@ -553,8 +553,14 @@ func LatestImageTagEvent(stream *ImageStream, imageID string) (string, *TagEvent
 			continue
 		}
 		for i, event := range events.Items {
-			if event.Image != imageID {
-				continue
+			if d, err := digest.ParseDigest(event.Image); err == nil {
+				if !strings.HasPrefix(d.Hex(), imageID) && !strings.HasPrefix(event.Image, imageID) {
```
Comment (pweil-) [inline pkg/image/api/helper.go]: we use this same digest and `event.Image` logic (and the item below) in the `ResolveImageId` in this same file.  Time for a helper method?

File: pkg/image/api/helper.go [hunk 2]
```diff
@@ -642,8 +648,10 @@ func ResolveLatestTaggedImage(stream *ImageStream, tag string) (string, bool) {
 func DockerImageReferenceForImage(stream *ImageStream, imageID string) (string, bool) {
 	tag, event := LatestImageTagEvent(stream, imageID)
 	if len(tag) == 0 {
+		glog.Infof("no tag")
```
Comment (pweil-) [inline pkg/image/api/helper.go]: are these two log statements useful?  Print out the stream and imageID maybe?

Comment (pweil-) [inline pkg/image/api/helper.go]: also, are they Info level or debugging level type?

**Comment by pweil-** [comment]
@legionus @miminar @soltysh

File: pkg/image/api/helper.go (hunk 2 above)
Comment (mfojtik) [inline pkg/image/api/helper.go]: sry, debugging left over.

File: pkg/image/api/helper.go (hunk 1 above)
Comment (mfojtik) [inline pkg/image/api/helper.go]: not sure if helper is possible... those two seems to have inverted logic.

**Comment by pweil-** [comment]
also per our retro discussion...where dem tests at?

**Comment by mfojtik** [comment]
@pweil- there is a `TestDockerImageReferenceForImage`, but you're right, I should add the non-sha256 format there to make sure both works.

File: pkg/image/api/helper.go (hunk 1 above)
Comment (pweil-) [inline pkg/image/api/helper.go]: that's what the `!` is for :smile: `DigestOrImageMatch()` vs `!DigestOrImageMatch()`

Comment (mfojtik) [inline pkg/image/api/helper.go]: ok, I gotcha and I was wrong :-)

**Comment by mfojtik** [comment]
@pweil- now with test and helper :-)

**Comment by mfojtik** [comment]
[test]

**Comment by pweil-** [comment]
LGTM

File: pkg/image/api/helper.go [hunk 3]
```diff
@@ -553,10 +553,7 @@ func LatestImageTagEvent(stream *ImageStream, imageID string) (string, *TagEvent
 			continue
 		}
 		for i, event := range events.Items {
-			if event.Image != imageID {
-				continue
-			}
-			if latestTagEvent == nil || (latestTagEvent != nil && event.Created.After(latestTagEvent.Created.Time)) {
+			if digestOrImageMatch(event.Image, imageID) && (latestTagEvent == nil || (latestTagEvent != nil && event.Created.After(latestTagEvent.Created.Time))) {
```
Comment (soltysh) [inline pkg/image/api/helper.go]: This condition freaks me out 😨

Comment (pweil-) [inline pkg/image/api/helper.go]: could be reduced to this, right?

```
if digestOrImageMatch(event.Image, imageID) && (latestTagEvent == nil || event.Created.After(latestTagEvent.Created.Time))
```

Comment (soltysh) [inline pkg/image/api/helper.go]: Yup.

Comment (mfojtik) [inline pkg/image/api/helper.go]: yeah, i'm a bracket guy...

Comment (mfojtik) [inline pkg/image/api/helper.go]: @pweil- if latestTagEvent is nil, then the right part will panic(), or i'm drunk?

**Comment by mfojtik** [comment]
[merge]

**Pull request merged**